File: Homework6/Hw6/Calculator.fs
module Hw6.Calculator

open System

type CalculatorOperation =
    | Plus = 0
    | Minus = 1
    | Multiply = 2
    | Divide = 3

[<Literal>]
let plus = "Plus"

[<Literal>]
let minus = "Minus"

[<Literal>]
let multiply = "Multiply"

[<Literal>]
let divide = "Divide"

[<System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage>]
let inline calculate value1 operation value2 : 'a =
    match operation with
    | CalculatorOperation.Plus -> value1 + value2
    | CalculatorOperation.Minus -> value1 - value2
    | CalculatorOperation.Divide -> value1 / value2
    | CalculatorOperation.Multiply -> value1 * value2
    | _ -> raise (InvalidOperationException($"{operation} is not a operation."))

File: Homework8/Hw8/Calculator/Calculator.cs
namespace Hw8.Calculator;

public class Calculator : ICalculator
{
    public double Calculate(double value1, Operation operation, double value2)
    {
        return operation switch
        {
            Operation.Plus => Plus(value1, value2),
            Operation.Minus => Minus(value1, value2),
            Operation.Multiply => Multiply(value1, value2),
            Operation.Divide => Divide(value1, value2),
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public double Plus(double val1, double val2)
    {
        return val1 + val2;
    }

    public double Minus(double val1, double val2)
    {
        return val1 - val2;
    }

    public double Multiply(double val1, double val2)
    {
        return val1 * val2;
    }

    public double Divide(double firstValue, double secondValue)
    {
        if (secondValue == 0) throw new InvalidOperationException();
        return firstValue / secondValue;
    }
}